packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -1059,6 +1069,8 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
                 allowedRequests.push(request);
             } else {
                 skipped.add(url);
+                const req = typeof request === 'string' ? { url: request } : request;
+                await this.onSkippedRequest?.(req as RequestOptions, 'robotsTxt');
```

Comment:
Awaiting on a user-defined handler in each iteration might slow the call down by a lot. Maybe we could postpone awaiting the calls until the end of `addRequests` and do a `Promise.all`?

Reply:
But that means you lose the order of those events, I would rather await them one by one. Maybe it's not really important to stay in order, but it feels like a nice thing to do.

> by a lot

Well, it depends on what is inside the callback. You are right this could add a lot if we go with the error dataset pushes or similar thing over the network.

Reply:
I mean, all this can happen in parallel with other tasks (`AutoscaledPool`), so some degree of scrambling is inevitable. I wouldn't mind scrambling it a bit more.

Reply:
yeah, good point